### PR TITLE
Add support for classic injection

### LINKED ISSUE #29: Configure injection mode
## 🚀 Allow change Awilix injection mode

When using Awilix with Typescript, CLASSIC injection mode seems to be more desirable.

## Motivation

Awilix creator @jeffijoe [comment](https://github.com/jeffijoe/awilix/issues/89#issuecomment-386412693):

> I used to prefer the options object (which was the reason I created Awilix using Proxies to begin with), however the problem with incorrect parameter passing goes away with TypeScript, and so I've started preferring the classic approach personally. Doesn't hurt that it's faster, too, hehe.

## Example

```typescript
fastify.register(fastifyAwilixPlugin, {
  injectionMode: 'CLASSIC' // InjectionMode.CLASSIC
})
```

### PR DESCRIPTION
**Pull request by kibertoad**
I'm somewhat on a fence over how to implement this. It is possible to parametrize existing implementation, but that would either introduce additional closure level, or have conditional branches all over the code, both of which would impact performance of a hot path. I've opted for an option that both allows tree-shaking and also has zero impact over hot path. It does, however, come with code duplication. Additional opinions on how this should be handled would be very welcome.

Also I wonder if there is a nicer way to type a submodule other than just putting a separate TS file there.

fixes #29

### CODE REVIEW

**Comment by kibertoad** [comment]
@jm42 Could you please take a look?

**Comment by jm42** [comment]
LGTM

I have duplicated the code also but I removed `diContainer` as export. What its use beside BC?

Will use this instead.

**Comment by kibertoad** [comment]
@jm42 It was explicitly requested in the past and I personally find it useful as well. Considering that it's effectively a singleton, it provides convenient access for cases where you don't necessarily have access to either app or request, but still want to inject something, for an example if you are going for a stateless architecture, based on functions and not classes.

**Comment by jm42** [comment]
@kibertoad I see. For me that would be like using the container as a Service Locator, which might work in some cases where the key to look for is dynamic. Anyway, I don't mind being created on import instead on plugin register.

File: lib/classic/index.js [hunk 1]
```diff
@@ -0,0 +1,43 @@
+const awilix = require('awilix')
```
Comment (Eomm) [inline lib/classic/index.js:1]:
```suggestion
'use strict'

const awilix = require('awilix')
```

File: lib/classic/index.js [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+const fp = require('fastify-plugin')
+
+const diContainer = awilix.createContainer({
```
Comment (Eomm) [inline lib/classic/index.js:4]: Could it give issue if the plugin is registered twice on different contexts?

```
fastify.register(async (instance)=>{
  instance.register(this plugin)
})
fastify.register(async (instance)=>{
  instance.register(this plugin)
})
```

File: lib/classic/index.js [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+  fastify.decorateRequest('diScope', null)
+
+  fastify.addHook('onRequest', (request, reply, done) => {
+    request.diScope = diContainer.createScope()
+    done()
+  })
+
+  if (disposeOnClose) {
+    fastify.addHook('onClose', (instance, done) => {
+      return diContainer.dispose().then(done)
```
Comment (Eomm) [inline lib/classic/index.js:22]:
```suggestion
      return diContainer.dispose().then(done, done)
```

File: lib/classic/index.js [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+    })
+  }
+
+  if (disposeOnResponse) {
+    fastify.addHook('onResponse', (request, reply, done) => {
+      return request.diScope.dispose().then(done)
```
Comment (Eomm) [inline lib/classic/index.js:28]:
```suggestion
      return request.diScope.dispose().then(done, done)
```

File: lib/index.js [hunk 5]
```diff
@@ -0,0 +1,6 @@
+const { diContainer, fastifyAwilixPlugin } = require('./fastifyAwilixPlugin')
```
Comment (Eomm) [inline lib/index.js:1]:
```suggestion
'use strict'

const { diContainer, fastifyAwilixPlugin } = require('./fastifyAwilixPlugin')
```

File: lib/classic/index.js (hunk 2 above)
Comment (kibertoad) [reply lib/classic/index.js:6]: Fair point. I will expose an option to pass user's own container for such cases.

File: lib/classic/index.js (hunk 3 above)
Comment (kibertoad) [reply lib/classic/index.js:24]: Why the second param?

Comment (jm42) [reply lib/classic/index.js:24]: I also use catch:

```ts
    fastify.addHook('onClose', (instance, done) =>
      container
        .dispose()
        .then(() => done())
        .catch(done)
    )
```

**Pull request merged**